vmm/src/vmm_config/mod.rs

```diff
@@ -16,3 +19,62 @@ pub mod net;
 #[cfg(feature = "vsock")]
 /// Wrapper for configuring the vsock devices attached to the microVM.
 pub mod vsock;
+
+// TODO: Migrate the VMM public-facing code (i.e. interface) to use stateless structures,
+// for receiving data/args, such as the below `RateLimiterConfig` and `TokenBucketConfig`.
+// Also todo: find a better suffix than `Config`; it should illustrate the static nature
+// of the enclosed data.
+// Currently, data is passed around using live/stateful objects. Switching to static/stateless
+// objects will simplify both the ownership model and serialization.
+// Public access would then be more tightly regulated via `VmmAction`s, consisting of tuples like
+// (entry-point-into-VMM-logic, stateless-args-structure).
+
+/// A public-facing, stateless structure, holding all the data we need to create a TokenBucket
+/// (live) object.
+#[derive(Clone, Copy, Debug, Default, Deserialize)]
+pub struct TokenBucketConfig {
+    /// See TokenBucket::size.
+    pub size: u64,
+    /// See TokenBucket::one_time_burst.
+    pub one_time_burst: Option<u64>,
+    /// See TokenBucket::refill_time.
+    pub refill_time: u64,
+}
+
+impl TokenBucketConfig {
+    /// Convert the stateless `self` into a live `TokenBucket` object.
+    ///
+    /// This would look nicer if we were to implement `Into<TokenBucket>` (or, more generally,
+    /// `Into<LiveCounterpart>`), but some constructors may fail, and until `TryInto` makes into
+    /// the stable channel, we'll settle for the unenforceable convention of implementing
+    /// a `into_<live_counterpart>()` member for these stateless structures.
+    pub fn into_token_bucket(self) -> TokenBucket {
+        TokenBucket::new(self.size, self.one_time_burst, self.refill_time)
+    }
+}
+
+/// A public-facing, stateless structure, holding all the data we need to create a RateLimiter
+/// (live) object.
+#[derive(Clone, Copy, Debug, Default, Deserialize)]
+pub struct RateLimiterConfig {
```

Comment:
With the `PatchRateLimiters` and `NetworkInterfaceUpdateConfig` this makes 4 structs that we use for rate limiters. Is there no way to squash them into fewer?

Reply:
Yes, we still have some clean up to do there. But I don't agree with merging functionally different, stateful objects into a single struct, even if they happen to share some data fields. Because static data definitions are easy to manage, while custom/branching logic isn't.